.github/workflows/tag-publish.yaml

```diff
@@ -1,10 +1,170 @@
+    permissions:
+      id-token: write
+    env:
+      ARTIFACT_TAG: ${{ inputs.artifact-tag || github.event.release.tag_name }}
+      LOCAL_ARTIFACTS_PATH: /tmp/artifacts
+      LOCAL_HELM_REPO_PATH: /tmp/helm-repo
+      LOCAL_TERRAFORM_REGISTRY_PATH: /tmp/terraform-registry
+      ENVIRONMENT_NAME: ${{ inputs.environment || 'release-prod' }}
     steps:
-      - run: echo "Placeholder"
+      - name: Validate artifact tag
+        env:
+          SEMVER_REGEX: ^v?(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)(?:-(?:(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?:[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$
+        run: |
+          (echo "$ARTIFACT_TAG" | grep -qP "$SEMVER_REGEX") || (echo "The artifact tag $ARTIFACT_TAG is not a valid semver-coerced value"; exit 1)
+      - name: Validate environment
+        run: |
+          (echo "$ENVIRONMENT_NAME" | grep -qP '^publish-(prod|stage)$') || (echo "This workflow may only be ran from publishing environments"; exit 1)
+      - name: Checkout repo
+        uses: actions/checkout@v4
+        with:
+          ref: ${{ github.workflow_sha }} # Publishing tooling should all be the same version
+      - name: Setup tooling Go
+        uses: actions/setup-go@v5
+        with:
+          go-version-file: "./tooling/go.mod"
+          cache-dependency-path: "./tooling/go.sum"
+          check-latest: true
+      # Sync artifacts down (required by everything)
+      - name: Assume AWS role for uploading the artifacts
+        uses: aws-actions/configure-aws-credentials@010d0da01d0b5a38af31e9c3470dbfdabdecca3a # v4.0.1
+        with:
+          role-skip-session-tagging: true
+          aws-region: us-west-2
+          role-to-assume: ${{ vars.ARTIFACT_DOWNLOAD_AWS_ROLE }}
+          role-session-name: "artifact-download-${{ github.run_number }}"
+          role-duration-seconds: 900
+      - name: Download artifacts from S3
+        env:
+          ARTIFACT_BUCKET: ${{ vars.ARTIFACT_SOURCE_BUCKET }}
+        run: aws s3 cp "s3://$ARTIFACT_BUCKET/teleport-plugins/tag/$ARTIFACT_TAG/" "$LOCAL_ARTIFACTS_PATH" --recursive
+      # Binary artifact promotion
+      - name: Assume AWS role for uploading the artifacts
+        # This step is only supported in production as there is no staging version of Houston
+        if: ${{ env.ENVIRONMENT_NAME == 'publish-prod' }}
+        uses: aws-actions/configure-aws-credentials@010d0da01d0b5a38af31e9c3470dbfdabdecca3a # v4.0.1
+        with:
+          role-skip-session-tagging: true
+          aws-region: us-west-2
+          role-to-assume: ${{ vars.HOUSTON_UPLOAD_AWS_ROLE }}
+          role-session-name: "houston-upload-${{ github.run_number }}"
+          role-duration-seconds: 900
+      - name: Upload artifacts to Houston
+        # This step is only supported in production as there is no staging version of Houston
+        if: ${{ env.ENVIRONMENT_NAME == 'publish-prod' }}
```

Comment:
perhaps instead just test that `$HOUSTON_BUCKET` is empty and exit with an error, so we can plug in a staging environment if we need to and not modify the workflow?

Reply:
> so we can plug in a staging environment if we need to and not modify the workflow

I don't think that this is reasonably likely to occur. I'm wrapping up some unrelated work right now, but once that's done my next order of business is to remove Houston entirely, starting with migrating Teleport Plugins off of it.